File: 10-regular-expression-matching/10-regular-expression-matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        int m = s.length(), n = p.length();
        boolean[][] dp = new boolean[m + 1][n + 1];
        
        s = ' ' + s;
        p = ' ' + p;
        dp[0][0] = true;
        for(int i = 0; i <= m; i++) {
            for(int j = 1; j <= n; j++) {
                if(j + 1 <= n && p.charAt(j + 1) == '*') continue;
                if(i != 0 && p.charAt(j) != '*') {
                    dp[i][j] = dp[i - 1][j - 1] && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.');
                } else if(p.charAt(j) == '*') {
                    dp[i][j] = dp[i][j - 2] || (i != 0 && dp[i - 1][j] && (s.charAt(i) == p.charAt(j - 1) || p.charAt(j - 1) == '.'));
                }
            }
        }
        return dp[m][n];
    }
}

File: 10-regular-expression-matching/10-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        m, n = len(s), len(p)
        dp = [[False]*(n + 1) for _ in range(m + 1)]
        
        s= ' '+ s
        p = ' ' + p
        dp[0][0] = True
        
        for i in range(m + 1):
            for j in range(1, n + 1):
                #特判
                if j + 1 <= n and p[j + 1] == '*':
                    continue
                if p[j] != '*':
                    if dp[i - 1][j - 1] and (s[i] == p[j] or p[j] == '.'):
                        dp[i][j] = True
                else:
                    dp[i][j] = dp[i][j - 2] or (i and dp[i -1][j] and (s[i] == p[j - 1] or p[j - 1] == '.'))
        return dp[m][n]